File: src/main/java/dev/vrba/secrethitler/engine/state/PlayerState.java
package dev.vrba.secrethitler.engine.state;

import dev.vrba.secrethitler.engine.Party;
import dev.vrba.secrethitler.engine.Role;
import dev.vrba.secrethitler.engine.claims.Claim;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.With;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Map;
import java.util.UUID;

@With
@Getter
@AllArgsConstructor
public class PlayerState {

    private final @NotNull UUID id;

    private final boolean alive;

    private final @NotNull Role role;

    private final @NotNull Map<UUID, Role> rolesKnowledge;

    private final @NotNull List<UUID> availablePlayerTargets;

    private final @NotNull List<Party> availablePolicies;

    private final @NotNull List<Claim> availableClaims;

    private final @NotNull List<Party> peekedPolicies;

    private final @Nullable Party investigatedPartyMembership;

    public @NotNull PublicPlayerState toPublicState() {
        return this.toPublicState(true);
    }

    public @NotNull PublicPlayerState toPublicState(boolean maskRole) {
        return new PublicPlayerState(
                this.id,
                this.alive,
                maskRole ? Role.HIDDEN : this.role
        );
    }
}

File: src/main/kotlin/dev/vrba/secrethitler/engine/state/PlayerState.kt
package dev.vrba.secrethitler.engine.state

import dev.vrba.secrethitler.engine.PartyMembership
import dev.vrba.secrethitler.engine.Policy
import dev.vrba.secrethitler.engine.Role
import dev.vrba.secrethitler.engine.claims.PartyMembershipClaim
import dev.vrba.secrethitler.engine.claims.PoliciesClaim
import java.util.*

/**
 * State that can be broadcasted to everybody without revealing personal knowledge (what is kept secret)
 */
data class PublicPlayerState(
    val id: UUID,
    val username: String,
    val alive: Boolean,
    val role: Role
)

data class PlayerState(
    val id: UUID,
    val username: String,
    val alive: Boolean,
    val role: Role,
    val rolesKnowledge: Map<UUID, Role> = emptyMap(),
    val availablePlayerTargets: List<UUID> = emptyList(),
    val availablePolicies: List<Policy> = emptyList(),
    val availablePoliciesClaims: List<PoliciesClaim> = emptyList(),
    val availablePartyMembershipClaims: List<PartyMembershipClaim> = emptyList(),
    val investigatedMemberships: Map<UUID, PartyMembership> = emptyMap(),
    val peekedPolicies: List<Policy> = emptyList()
) {
    fun public(maskRoles: Boolean = true): PublicPlayerState =
        PublicPlayerState(
            id,
            username,
            alive,
            if (maskRoles) Role.Hidden else role
        )
}